state/clouduser.go

```diff
@@ -92,3 +98,118 @@ func (st *State) RemoveCloudAccess(cloud string, user names.UserTag) error {
 	err := st.db().Run(buildTxn)
 	return errors.Trace(err)
 }
+
+// CloudInfo describes interesting information for a given cloud.
+type CloudInfo struct {
+	cloud.Cloud
+
+	// Access is the access level the supplied user has on this cloud.
+	Access permission.Access
+}
+
+// CloudsForUser returns details including access level of clouds which can
+// be seen by the specified user, or all users if the caller is a superuser.
+func (st *State) CloudsForUser(user names.UserTag, all bool) ([]CloudInfo, error) {
+	// We only treat the user as a superuser if they pass --all
+	isControllerSuperuser := false
+	if all {
+		var err error
+		isControllerSuperuser, err = st.isUserSuperuser(user)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	cloudQuery, closer, err := st.cloudQueryForUser(user, isControllerSuperuser)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer closer()
+	var cloudDocs []cloudDoc
+	if err := cloudQuery.All(&cloudDocs); err != nil {
+		return nil, errors.Trace(err)
+	}
+	result := make([]CloudInfo, len(cloudDocs))
+	for i, c := range cloudDocs {
+		result[i] = CloudInfo{
+			Cloud: c.toCloud(),
+		}
+	}
+	if err := st.fillInCloudUserAccess(user, result); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return result, nil
+}
+
+// cloudQueryForUser returns a query returning the details of, and a helper function
+// which returns the access level for, clouds that a given user can see.
+func (st *State) cloudQueryForUser(user names.UserTag, isSuperuser bool) (mongo.Query, SessionCloser, error) {
+	var cloudQuery mongo.Query
+	clouds, closer := st.db().GetCollection(cloudsC)
+
+	if isSuperuser {
+		// Fast path, we just return all the clouds.
+		cloudQuery = clouds.Find(nil)
```

Comment:
I think it would make more sense to make this function just about getting a specific user's clouds. I'd pull the collection, closer and this superuser path up to CloudForUser.

Reply:
done